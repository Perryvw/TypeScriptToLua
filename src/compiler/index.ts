export { Chunk } from "./chunk";
export * from "./compiler";
export { Module } from "./module";
export { Plugin } from "./plugins";
export * from "./transpile";
