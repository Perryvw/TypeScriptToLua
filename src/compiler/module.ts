import { SourceNode } from "source-map";
import * as ts from "typescript";
import { escapeString, unescapeLuaString } from "../LuaPrinter";

/**
 * Source code of a single input Lua file.
 * May be constructed from transpiled `.ts` source files, or from real `.lua` files.
 */
export interface Module {
    fileName: string;
    isBuilt: boolean;
    source: SourceNode;
    sourceFiles?: ts.SourceFile[];
}

export type ModuleDependencyResolver = (request: string, position: ts.ReadonlyTextRange) => string | { error: string };

export function buildModule(module: Module, dependencyResolver: ModuleDependencyResolver) {
    if (module.isBuilt) return;
    module.isBuilt = true;
    replaceResolveRequests(module.source, dependencyResolver);
}

function replaceResolveRequests(rootNode: SourceNode, dependencyResolver: ModuleDependencyResolver) {
    let currentPosition = 0;

    function replaceInString(source: string) {
        const matches = source.matchAll(/__TS__Resolve\((".*?")\)/g);
        for (const match of [...matches].reverse()) {
            const request = unescapeLuaString(match[1]);
            const pos = currentPosition + match.index!;
            const end = pos + match[0].length;
            const result = dependencyResolver(request, { pos, end });
            const replacement =
                typeof result === "string"
                    ? escapeString(result)
                    : `--[[ ${request} ]] error(${escapeString(result.error)})`;

            source = source.slice(0, match.index) + replacement + source.slice(end);
        }

        return source;
    }

    function walkSourceNode(node: SourceNode, parent: SourceNode) {
        for (const child of node.children as Array<SourceNode | string>) {
            if (typeof child === "object") {
                walkSourceNode(child, node);
            } else {
                if (child.includes("__TS__Resolve")) {
                    parent.children = [replaceInString(parent.toString()) as any];
                }

                currentPosition += child.length;
            }
        }
    }

    walkSourceNode(rootNode, rootNode);
}
