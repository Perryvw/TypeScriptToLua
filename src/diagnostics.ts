import * as ts from "typescript";
import { TranspileError } from "./TranspileError";

export const transpileError = (error: TranspileError): ts.Diagnostic => ({
    file: error.node.getSourceFile(),
    start: error.node.getStart(),
    length: error.node.getWidth(),
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: error.message,
});

export const tstlOptionsAreMovingToTheTstlObject = (tstl: Record<string, any>): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Warning,
    code: 0,
    source: "typescript-to-lua",
    messageText:
        'TSTL options are moving to the "tstl" object. Adjust your tsconfig to look like\n' +
        `"tstl": ${JSON.stringify(tstl, undefined, 4)}`,
});

export const toLoadTransformerItShouldBeTranspiled = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `To load "${transform}" transformer it should be transpiled or "ts-node" should be installed`,
});

export const couldNotResolveTransformerFrom = (transform: string, base: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `Could not resolve "${transform}" transformer from "${base}".`,
});

export const transformerShouldHaveAExport = (transform: string, importName: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should have a "${importName}" export`,
});

export const transformerShouldBeATsTransformerFactory = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should be a ts.TransformerFactory or an object with ts.TransformerFactory values`,
});

export const watchErrorSummary = (errorCount: number): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Message,
    code: errorCount === 1 ? 6193 : 6194,
    messageText:
        errorCount === 1
            ? "Found 1 error. Watching for file changes."
            : `Found ${errorCount} errors. Watching for file changes.`,
});

const createCommandLineError = <Args extends any[]>(code: number, getMessage: (...args: Args) => string) => (
    ...args: Args
): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code,
    messageText: getMessage(...args),
});

export const unknownCompilerOption = createCommandLineError(
    5023,
    (name: string) => `Unknown compiler option '${name}'.`
);

export const compilerOptionRequiresAValueOfType = createCommandLineError(
    5024,
    (name: string, type: string) => `Compiler option '${name}' requires a value of type ${type}.`
);

export const optionProjectCannotBeMixedWithSourceFilesOnACommandLine = createCommandLineError(
    5042,
    () => "Option 'project' cannot be mixed with source files on a command line."
);

export const cannotFindATsconfigJsonAtTheSpecifiedDirectory = createCommandLineError(
    5057,
    (dir: string) => `Cannot find a tsconfig.json file at the specified directory: '${dir}'.`
);

export const theSpecifiedPathDoesNotExist = createCommandLineError(
    5058,
    (dir: string) => `The specified path does not exist: '${dir}'.`
);

export const compilerOptionExpectsAnArgument = createCommandLineError(
    6044,
    (name: string) => `Compiler option '${name}' expects an argument.`
);

export const argumentForOptionMustBe = createCommandLineError(
    6046,
    (name: string, values: string) => `Argument for '${name}' option must be: ${values}.`
);

export const optionCanOnlyBeSpecifiedInTsconfigJsonFile = createCommandLineError(
    6064,
    (name: string) => `Option '${name}' can only be specified in 'tsconfig.json' file.`
);

export const optionBuildMustBeFirstCommandLineArgument = createCommandLineError(
    6369,
    () => "Option '--build' must be the first command line argument."
);
