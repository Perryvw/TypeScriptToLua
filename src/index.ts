import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileWithSystem } from "./CommandLineParser";
import { CompilerOptions } from "./CompilerOptions";
import { getTranspileOutput, TranspilationResult, TranspiledFile } from "./Transpile";

export { parseCommandLine, ParsedCommandLine, updateParsedConfigFile } from "./CommandLineParser";
export { CompilerOptions, LuaLibImportKind, LuaTarget } from "./CompilerOptions";
export * from "./Emit";
export * from "./LuaAST";
export { LuaLibFeature } from "./LuaLib";
export { LuaPrinter } from "./LuaPrinter";
export { LuaTransformer } from "./LuaTransformer";
export * from "./Transpile";

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {}
): TranspilationResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { transpiledFiles, diagnostics: [...allDiagnostics] };
}

export function transpileProject(fileName: string, options?: CompilerOptions): TranspilationResult {
    const parseResult = parseConfigFileWithSystem(fileName, options);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, transpiledFiles: new Map() };
    }

    return transpileFiles(parseResult.fileNames, parseResult.options);
}

const libCache: { [key: string]: ts.SourceFile } = {};

/** @internal */
export function createVirtualProgram(
    input: Record<string, string>,
    options?: CompilerOptions
): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile: () => {},

        getSourceFile: filename => {
            if (filename in input) {
                return ts.createSourceFile(
                    filename,
                    input[filename],
                    ts.ScriptTarget.Latest,
                    false
                );
            }

            if (filename.startsWith("lib.")) {
                if (libCache[filename]) return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");

                libCache[filename] = ts.createSourceFile(
                    filename,
                    content,
                    ts.ScriptTarget.Latest,
                    false
                );

                return libCache[filename];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface VirtualProgramResult {
    file: TranspiledFile;
    diagnostics: ts.Diagnostic[];
}

export function transpileString(
    content: string,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const program = createVirtualProgram({ "main.ts": content }, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get("main.ts"), diagnostics: [...allDiagnostics] };
}

export function transpileVirtualProgram(
    files: Record<string, string>,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const mainFileName = Object.keys(files).find(x => /\bmain\.[a-z]+$/.test(x));
    if (mainFileName === undefined) {
        throw new Error('Virtual program should have a file named "main"');
    }

    const program = createVirtualProgram(files, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get(mainFileName), diagnostics: [...allDiagnostics] };
}
