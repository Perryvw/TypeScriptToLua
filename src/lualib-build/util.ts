import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
