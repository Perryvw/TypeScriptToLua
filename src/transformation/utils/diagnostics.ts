import * as ts from "typescript";

const createDiagnosticFactory = <TArgs extends any[]>(
    message: string | ((...args: TArgs) => string),
    category = ts.DiagnosticCategory.Error
) => (node: ts.Node, ...args: TArgs): ts.Diagnostic => ({
    file: node.getSourceFile(),
    start: node.getStart(),
    length: node.getWidth(),
    category,
    code: 0,
    source: "typescript-to-lua",
    messageText: typeof message === "string" ? message : message(...args),
});

export const forbiddenForIn = createDiagnosticFactory(`Iterating over arrays with 'for ... in' is not allowed.`);

export const invalidForRangeCall = createDiagnosticFactory((message: string) => `Invalid @forRange call: ${message}.`);
