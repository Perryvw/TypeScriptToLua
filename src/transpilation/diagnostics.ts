import * as ts from "typescript";

export const toLoadTransformerItShouldBeTranspiled = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `To load "${transform}" transformer it should be transpiled or "ts-node" should be installed`,
});

export const couldNotResolveTransformerFrom = (transform: string, base: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `Could not resolve "${transform}" transformer from "${base}".`,
});

export const transformerShouldHaveAExport = (transform: string, importName: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should have a "${importName}" export`,
});

export const transformerShouldBeATsTransformerFactory = (transform: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `"${transform}" transformer should be a ts.TransformerFactory or an object with ts.TransformerFactory values`,
});

export const couldNotFindBundleEntryPoint = (entryPoint: string): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Error,
    code: 0,
    source: "typescript-to-lua",
    messageText: `Could not find bundle entry point '${entryPoint}'. It should be a file in the project.`,
});
