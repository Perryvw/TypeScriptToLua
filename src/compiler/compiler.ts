import { FileSystem } from "enhanced-resolve";
import * as ts from "typescript";
import { Compilation } from "./compilation";
import { Module } from "./module";
import { Plugin } from "./plugins";
import { emitProgramModules, TranspileOptions } from "./transpile";

export interface CompilerHost extends Pick<ts.System, "getCurrentDirectory" | "readFile" | "writeFile"> {
    resolutionFileSystem?: FileSystem;
}

export interface EmitOptions extends TranspileOptions {
    writeFile?: ts.WriteFileCallback;
    plugins?: Plugin[];
}

export interface EmitResult {
    emitSkipped: boolean;
    diagnostics: readonly ts.Diagnostic[];
}

export class Compiler {
    public host: CompilerHost;
    constructor({ host = ts.sys }: { host?: CompilerHost } = {}) {
        this.host = host;
    }

    public emit(emitOptions: EmitOptions): EmitResult {
        const { program, writeFile = this.host.writeFile } = emitOptions;
        const compilation = new Compilation(this, program, emitOptions.plugins ?? []);
        const { options } = compilation;

        // Clean-up module cache from modules created from deleted source files, so they won't be hit during resolution
        const programSourceFiles = program.getSourceFiles();
        const validModulesInCache = [...this.moduleCache.values()].filter(
            module => !module.sourceFiles?.some(sourceFile => !programSourceFiles.includes(sourceFile))
        );
        this.moduleCache.clear();
        validModulesInCache.forEach(m => this.addModuleToCache(m));

        emitProgramModules(compilation, writeFile, emitOptions);
        if (options.noEmit || (options.noEmitOnError && compilation.diagnostics.length > 0)) {
            return { diagnostics: compilation.diagnostics, emitSkipped: true };
        }

        compilation.emit(writeFile);

        return { diagnostics: compilation.diagnostics, emitSkipped: false };
    }

    // A module cache that persists between compilations, so it can generate ids for these modules,
    // without re-reading files every time
    private moduleCache = new Map<string, Module>();

    public addModuleToCache(module: Module) {
        this.moduleCache.set(module.fileName, module);
    }

    public findModuleInCache(fileName: string) {
        return this.moduleCache.get(fileName);
    }
}
