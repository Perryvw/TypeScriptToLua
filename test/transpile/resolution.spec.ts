import * as tstl from "../../src";
import { createResolutionErrorDiagnostic } from "../../src/compiler/diagnostics";
import * as util from "../util";

const requireRegex = /require\("(.*?)"\)/;
const expectToRequire = (expected: string): util.TapCallback => builder => {
    const [, requiredPath] = builder.getMainLuaCodeChunk().match(requireRegex) ?? [];
    expect(requiredPath).toBe(expected);
};

const expectModuleTableToMatchSnapshot: util.TapCallback = builder => {
    builder.expectToHaveNoDiagnostics();
    const moduleTable = builder.getMainLuaCodeChunk().match(/(?<=\n)____modules = {\n(.+)\n}\nreturn [^\n]+$/s)?.[1];
    expect(moduleTable).not.toBeUndefined();
    expect(moduleTable).toMatchSnapshot("modules");
};

test("sibling file", () => {
    util.testBundle`
        import "./foo";
    `
        .addExtraFile("foo.ts", "")
        .tap(expectModuleTableToMatchSnapshot);
});

test("file in a sibling directory", () => {
    util.testBundle`
        import "./foo/bar.ts";
    `
        .addExtraFile("foo/bar.ts", "")
        .tap(expectModuleTableToMatchSnapshot);
});

test("sibling directory index", () => {
    util.testBundle`
        import "./foo";
    `
        .addExtraFile("foo/index.ts", "")
        .tap(expectModuleTableToMatchSnapshot);
});

test("current directory index", () => {
    util.testBundle`
        import ".";
    `
        .addExtraFile("index.ts", "")
        .tap(expectModuleTableToMatchSnapshot);
});

test("rootDir inference", () => {
    util.testBundle`
        import "./module";
    `
        .setMainFileName("src/main.ts")
        .addExtraFile("src/module.ts", "")
        .tap(expectModuleTableToMatchSnapshot);
});

test("entry point in nested directory", () => {
    util.testBundle`
        import "./module";
    `
        .setMainFileName("src/main.ts")
        .addExtraFile("src/module.ts", "")
        .setOptions({ rootDir: "src" })
        .tap(expectModuleTableToMatchSnapshot);
});

describe("resolution out of rootDir", () => {
    test(".lua file", () => {
        util.testBundle`
            export { value } from "../module";
        `
            .setOptions({ rootDir: "src" })
            .setMainFileName("src/main.ts")
            .addExtraFile("module.d.ts", "export declare const value: boolean;")
            .addRawFile("module.lua", "return { value = true }")

            .tap(expectToRequire("_.module"))
            .expectToEqual({ value: true });
    });

    test(".ts file", () => {
        util.testBundle`
            export { value } from "../module";
        `
            .setOptions({ rootDir: "src" })
            .setMainFileName("src/main.ts")
            .addExtraFile("module.ts", "export const value = true;")

            .tap(expectToRequire("_.module"))
            .expectDiagnosticsToMatchSnapshot([6059], true)
            .expectToEqual({ value: true });
    });
});

describe("package resolution", () => {
    test("without package.json", () => {
        util.testBundle`
            export { value } from 'lib';
        `
            .addExtraFile("node_modules/lib/index.d.ts", "export const value: boolean;")
            .addRawFile("node_modules/lib/index.lua", "return { value = true }")
            .tap(expectModuleTableToMatchSnapshot)
            .expectToEqual({ value: true });
    });

    test("package.json with exports", () => {
        util.testBundle`
            export { value } from 'lib';
        `
            .addExtraFile("node_modules/lib/index.d.ts", "export const value: boolean;")
            .addRawFile("node_modules/lib/dist/index.lua", "return { value = true }")
            .addRawFile("node_modules/lib/package.json", JSON.stringify({ exports: { lua: "./dist/index.lua" } }))
            .tap(expectModuleTableToMatchSnapshot)
            .expectToEqual({ value: true });
    });

    // https://github.com/webpack/enhanced-resolve/issues/256
    test.skip("package.json with directory exports", () => {
        util.testBundle`
            export { value } from 'lib/foo';
        `
            .addExtraFile("node_modules/lib/index.d.ts", 'declare module "lib/foo" { export const value: boolean; }')
            .addRawFile("node_modules/lib/dist/foo.lua", "return { value = true }")
            .addRawFile("node_modules/lib/package.json", JSON.stringify({ exports: { "./": { lua: "./dist/" } } }))
            .tap(expectModuleTableToMatchSnapshot)
            .expectToEqual({ value: true });
    });

    test("package.json with versioned exports", () => {
        util.testBundle`
            export { value } from 'lib';
        `
            .addExtraFile("node_modules/lib/index.d.ts", "export const value: boolean;")
            .addRawFile("node_modules/lib/dist/5.3.lua", 'return { value = "5.3" }')
            .addRawFile("node_modules/lib/dist/jit.lua", 'return { value = "jit" }')
            .addRawFile(
                "node_modules/lib/package.json",
                JSON.stringify({ exports: { "lua:5.3": "./dist/5.3.lua", "lua:jit": "./dist/jit.lua" } })
            )
            .tap(expectModuleTableToMatchSnapshot)
            .expectToEqual({ value: "5.3" });
    });

    test("package with dependencies", () => {
        util.testBundle`
            export { value } from 'lib';
        `
            .addExtraFile("node_modules/lib/index.d.ts", "export const value: boolean;")
            .addRawFile("node_modules/lib/index.lua", 'return require(__TS__Resolve("lib2"))')
            .addRawFile("node_modules/lib2/index.lua", "return { value = true }")
            .tap(expectModuleTableToMatchSnapshot)
            .expectToEqual({ value: true });
    });

    test.todo("symlink");
});

test("not transpiled script file error", () => {
    util.testBundle`
        declare function require(this: void, path: string): any;
        declare function __TS__Resolve(this: void, request: string): string;

        export const { value } = require(__TS__Resolve("./module"));
    `
        .addRawFile("module.ts", "export const value = true;")
        .expectDiagnosticsToMatchSnapshot([createResolutionErrorDiagnostic.code], true)
        .tap(expectModuleTableToMatchSnapshot)
        .expectToEqual(new util.ExecutionError("Resolved source file '/module.ts' is not a part of the project."));
});

test.each([
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {}
        `,
        mainCode: 'import "fake";',
    },
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {}
        `,
        mainCode: 'import * as fake from "fake"; fake;',
    },
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {
                export const x: number;
            }
        `,
        mainCode: 'import { x } from "fake"; x;',
    },
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {
                export const x: number;
            }

            declare module "fake" {
                export const y: number;
            }
        `,
        mainCode: 'import { y } from "fake"; y;',
    },
])("@noResolution annotation (%p)", ({ declarationStatement, mainCode }) => {
    util.testModule(mainCode)
        .setMainFileName("src/main.ts")
        .addExtraFile("module.d.ts", declarationStatement)
        .tap(expectToRequire("fake"));
});

describe('mode: "lib"', () => {
    test("doesn't fail on unresolvable requests", () => {
        util.testBundle`
            import "./module";
        `
            .addExtraFile("module.d.ts", "export const foo = true;")
            .setOptions({ mode: tstl.CompilerMode.Lib })
            .expectToHaveNoDiagnostics()
            .tap(expectModuleTableToMatchSnapshot);
    });

    test.todo('gets resolved with mode: "app" from different compilation');
});
