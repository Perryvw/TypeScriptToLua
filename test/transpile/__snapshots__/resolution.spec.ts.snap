// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`current directory index: modules 1`] = `
"[\\"index\\"] = function()

end,
[\\"main\\"] = function()
local ____exports = {}
require(\\"index\\")
return ____exports

end,"
`;

exports[`entry point in nested directory: modules 1`] = `
"[\\"module\\"] = function()

end,
[\\"main\\"] = function()
local ____exports = {}
require(\\"module\\")
return ____exports

end,"
`;

exports[`file in a sibling directory: modules 1`] = `
"[\\"foo.bar\\"] = function()

end,
[\\"main\\"] = function()
local ____exports = {}
require(\\"foo.bar\\")
return ____exports

end,"
`;

exports[`mode: "lib" doesn't fail on unresolvable requests: modules 1`] = `
"[\\"main\\"] = function()
local ____exports = {}
require(__TS__Resolve(\\"./module\\"))
return ____exports

end,"
`;

exports[`not transpiled script file error: diagnostics 1`] = `"/main.ts(2,22): error TSTL: Resolved source file '/module.ts' is not a part of the project."`;

exports[`not transpiled script file error: modules 1`] = `
"[\\"main\\"] = function()
local ____exports = {}
local ____ = require(--[[ ./module ]] error(\\"Resolved source file '/module.ts' is not a part of the project.\\"))
____exports.value = ____.value
return ____exports

end,"
`;

exports[`package resolution package with dependencies: modules 1`] = `
"[\\"main\\"] = function()
local ____exports = {}
do
    local ____lib = require(\\"node_modules.lib.index\\")
    local value = ____lib.value
    ____exports.value = value
end
return ____exports

end,
[\\"node_modules.lib.index\\"] = function()
return require(\\"node_modules.lib2.index\\")
end,
[\\"node_modules.lib2.index\\"] = function()
return { value = true }
end,"
`;

exports[`package resolution package.json with exports: modules 1`] = `
"[\\"main\\"] = function()
local ____exports = {}
do
    local ____lib = require(\\"node_modules.lib.dist.index\\")
    local value = ____lib.value
    ____exports.value = value
end
return ____exports

end,
[\\"node_modules.lib.dist.index\\"] = function()
return { value = true }
end,"
`;

exports[`package resolution package.json with versioned exports: modules 1`] = `
"[\\"main\\"] = function()
local ____exports = {}
do
    local ____lib = require(\\"node_modules.lib.dist.5__3\\")
    local value = ____lib.value
    ____exports.value = value
end
return ____exports

end,
[\\"node_modules.lib.dist.5__3\\"] = function()
return { value = \\"5.3\\" }
end,"
`;

exports[`package resolution without package.json: modules 1`] = `
"[\\"main\\"] = function()
local ____exports = {}
do
    local ____lib = require(\\"node_modules.lib.index\\")
    local value = ____lib.value
    ____exports.value = value
end
return ____exports

end,
[\\"node_modules.lib.index\\"] = function()
return { value = true }
end,"
`;

exports[`prioritize .lua files over external .ts: modules 1`] = `
"[\\"main\\"] = function()
local ____exports = {}
require(\\"foo\\")
return ____exports

end,
[\\"foo\\"] = function()

end,"
`;

exports[`prioritize internal .ts files over .lua: modules 1`] = `
"[\\"foo\\"] = function()

end,
[\\"main\\"] = function()
local ____exports = {}
require(\\"foo\\")
return ____exports

end,"
`;

exports[`resolution out of rootDir .ts file: diagnostics 1`] = `"src/main.ts(2,35): error TS6059: File 'module.ts' is not under 'rootDir' 'src'. 'rootDir' is expected to contain all source files."`;

exports[`rootDir inference: modules 1`] = `
"[\\"module\\"] = function()

end,
[\\"main\\"] = function()
local ____exports = {}
require(\\"module\\")
return ____exports

end,"
`;

exports[`sibling directory index: modules 1`] = `
"[\\"foo.index\\"] = function()

end,
[\\"main\\"] = function()
local ____exports = {}
require(\\"foo.index\\")
return ____exports

end,"
`;

exports[`sibling file: modules 1`] = `
"[\\"foo\\"] = function()

end,
[\\"main\\"] = function()
local ____exports = {}
require(\\"foo\\")
return ____exports

end,"
`;
